src/FixedRankMatrices.jl

```diff
@@ -0,0 +1,177 @@
+    if c != nothing
+        return c
+    end
+    if (size(v.U) != (M,k)) || (size(v.Vt) != (k,N)) || (size(v.M) != (k,k))
+        return DomainError(cat(size(v.U),size(v.M),size(v.Vt),dims=1), "The tangent vector $(v) is not a tangent vector to $(x) on the fixed rank matrices since the matrix dimensions to not fit (expected $([M,k,k,k,k,N])).")
+    end
+    if !isapprox(v.U'*x.U, zeros(k,k))
+        return DomainError(norm(v.U'*x.U-zeros(k,k)), "The tangent vector $(v) is not a tangent vector to $(x) on the fixed rank matrices since v.U'x.U is not zero. ")
+    end
+    if !isapprox(v.Vt*x.Vt', zeros(k,k))
+        return DomainError(norm(v.Vt*x.Vt-zeros(k,k)), "The tangent vector $(v) is not a tangent vector to $(x) on the fixed rank matrices since v.V'x.V is not zero.")
+    end
+end
+
+function inner(::FixedRankMatrices{M,N,k,T}, x::SVDMPoint, v::UMVTVector, w::UMVTVector) where {M,N,k,T}
+    return dot(v.U,w.U) + dot(v.M,w.M) + dot(v.Vt,w.Vt)
+end
+
+manifold_dimension(::FixedRankMatrices{M,N,k,T}) where {M,N,k,T <: Real} = (M+N-k)*k
+manifold_dimension(::FixedRankMatrices{M,N,k,T}) where {M,N,k,T <: Complex{U} where U} = 2*(M+N-k)*k
+
+@doc doc"""
+    project_tangent!(M,vto,x,A)
+
+project the matrix $A\in\mathbb R^{m,n}$ or a [`UMVTVector`](@ref) (e.g. from
+another tangent space) onto the tangent space at $x$, further
+decomposing the result into $v=UMV$, i.e. a [`UMVTVector`](@ref) following
+Section 3 in 
+> Bart Vandereycken: "Low-rank matrix completion by Riemannian Optimization,
+> SIAM Journal on Optiomoization, 23(2), pp. 1214–1236, 2013.
+> doi: [10.1137/110845768](https://doi.org/10.1137/110845768),
+> arXiv: [1209.3834](https://arxiv.org/abs/1209.3834).
+"""
+function project_tangent!(::FixedRankMatrices{M,N,k,T}, vto::UMVTVector, x::SVDMPoint, A::Matrix) where {M,N,k,T}
+    vto.M .= x.U * A * x.Vt'
+    vto.U .= A * x.Vt' - x.U
+    vto.Vt .= x.U' * A - x.U' * A * x.Vt' * x.Vt
+    return vto
+end
+project_tangent!(F::FixedRankMatrices{M,N,k,T}, vto::UMVTVector, x::SVDMPoint, v::UMVTVector) where {M,N,k,T} = project_tangent!(F,vto,x, v.U*v.M.v.Vt)
+
+@doc doc"""
+    retract!(M, y, x, v, ::PolarRetraction)
+
+compute an SVD-based retraction on the [`FixedRankMatrices`](@ref) manifold
+by computing
+````math
+    y = U_kS_kV_k^\mathrm{T},
+````
+where $U_kS_kV_k^\mathmr{T}$ is the shortened singular value decomposition $USV=x+v$,
+in the sense that $S_k$ is the diagonal matrix of size $k\times k$ with the $k$ largest
+singular values and $U$ and $V$ are shortened accordingly.
+"""
+function retract!(::FixedRankMatrices{M,N,k,T}, y::SVDMPoint, x::SVDMPoint, v::UMVTVector, ::PolarRetraction) where {M,N,k,T}
+    s = svd( x.U * Diagonal(x.S) * x.Vt + (x.U * v.M * x.Vt + v.U*x.Vt + v.U*v.Vt) )
+    y.U .= s.U[:,1:k]
+    y.S .= s.S[1:k]
+    y.Vt .= s.Vt[1:k,:]
+    return y
+end
+
+zero_tangent_vector(::FixedRankMatrices{m,n,k,T},x::SVDMPoint) where {m,n,k,T} = UMTVector(zeros(T,n,k), zeros(T,k,k), zeros(T,k,m))
```

Comment:
The mutating variant would also be nice

Reply:
Oh, it should actually be enough to do _just_ the mutating one, the other one is already generally given in `Manifolds`.